File: HashMap/930. Binary Subarrays With Sum/Solution.java
class Solution {
    public int numSubarraysWithSum(int[] A, int S) {
        int n = A.length;
        //key: prefix sum
        //value: count
        Map<Integer, Integer> map = new HashMap<>();
        map.put(0, 1);
        int count=0;
        int sum=0;
        for (int a:A){
            sum += a;
            if (map.containsKey(sum-S)){
                count+=map.get(sum-S);
            }
            map.put(sum, map.getOrDefault(sum,0)+1);
        }
        return count;
    }
}

File: HashMap/930. Binary Subarrays With Sum/Solution.py
class Solution(object):
    def numSubarraysWithSum(self, A, S):
        """
        :type A: List[int]
        :type S: int
        :rtype: int
        """
        prefix={}
        prefix[0]=1
        n = len(A)
        count=0
        cur_sum=0
        for a in A:
            cur_sum+=a
            remain = cur_sum-S
            if remain in prefix:
                count += prefix[remain]
            prefix[cur_sum] = prefix.get(cur_sum, 0)+1
        return count